fix docker inspect return unconsistent network settings  of created container and stopped container

coolljt0725 opened the pull request:
To make docker inspect return a consistent result of networksettings
for created container and stopped container, it's better to update
the network settings on container creating. Or else `docker inspect` a created 
container will not return the network settings.

<pre><code>$ docker create --name foo -ti busybox
4b5600da39e0350a3bb872f1d6a790443b327c87a2baa21300cf8e6abc09c85f
docker inspect foo
.
.
"NetworkSettings": {
        "Bridge": "",
        "SandboxID": "",
        "HairpinMode": false,
        "LinkLocalIPv6Address": "",
        "LinkLocalIPv6PrefixLen": 0,
        "Ports": null,
        "SandboxKey": "",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null,
        "EndpointID": "",
        "Gateway": "",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "IPAddress": "",
        "IPPrefixLen": 0,
        "IPv6Gateway": "",
        "MacAddress": "",
        "Networks": null
    }
</code></pre>

as you can see `Networks` filed is `null`, it should has the network `bridge`, but if start it and then stop it, then `docker inspect` will return `Networks`

<pre><code>$ docker start foo
foo
$ docker stop foo
foo
$ docker inspect foo
.
.
"Networks": {
            "bridge": {
                "EndpointID": "",
                "Gateway": "",
                "IPAddress": "",
                "IPPrefixLen": 0,
                "IPv6Gateway": "",
                "GlobalIPv6Address": "",
                "GlobalIPv6PrefixLen": 0,
                "MacAddress": ""
            }
        }
</code></pre>

It should have the consistent result.
Signed-off-by: Lei Jitang leijitang@huawei.com

thaJeztah commented:
Will (with this change) `Networks` show all networks that a container was connected to when running, or only "bridge"? i.e., if a container was *not* connected to "bridge" will it still appear in `Networks` when stopped? And the reverse; if a container was connected to multiple networks, will all networks show in `Networks` when stopped?

coolljt0725 commented:
@thaJeztah  This change will only affect the `created` container. 
>Will (with this change) Networks show all networks that a container was connected to when running, or only "bridge"

all networks,  and nothing to do with this change
>i.e., if a container was not connected to "bridge" will it still appear in Networks when stopped?

no, it will disappear, and also nothing to do with this change
>And the reverse; if a container was connected to multiple networks, will all networks show in 
Networks when stopped

yes, all networks will show, and also nothing to do with this change

cpuguy83 commented:
ping @aboch @mavenugo Just want to make sure you are good on this.

cpuguy83 commented:
Design review so the nework team can verify

mavenugo commented:
Design LGTM especially we need this in order to support multiple `--net` as discussed in https://github.com/docker/docker/pull/17796#issuecomment-162612624 (option #2).

But to clarify one of the questions that @thaJeztah asked 

>>    i.e., if a container was not connected to "bridge" will it still appear in Networks when stopped?

> no, it will disappear, and also nothing to do with this change

The answer depends on the case. If the container is NOT connected to "bridge" but to another network, then it will show up in the Networks when the container is stopped.

I have a few comments on the code though.

1. To avoid Code duplication, UpdateNetworkSettings must be reused from within `allocateNetwork`.

2. If we update the NetworkSettings in Create, can you please make sure it stays if the daemon is restarted ? (I believe we must write the config into the JSON file using `container.ToDiskLocking`

cpuguy83 commented:
Ok, moving to code review.
Thanks!

coolljt0725 commented:
@mavenugo  
>If we update the NetworkSettings in Create, can you please make sure it stays if the daemon is restarted ? (I believe we must write the config into the JSON file using container.ToDiskLocking

It will stays if the daemon is restarted, because after `daemon.UpdateNetworkSettings(container)` , 
it is `if err := container.ToDiskLocking();`
https://github.com/docker/docker/pull/18531/files#diff-72dc1842ed7b5a92b4e98f84456a2c0cR113

coolljt0725 commented:
@mavenugo  updated

mavenugo commented on daemon/container_operations_unix.go:
```diff
@@ -539,24 +562,10 @@ func (daemon *Daemon) allocateNetwork(container *container.Container) error {
 	updateSettings := false
 	if len(container.NetworkSettings.Networks) == 0 {
-		mode := container.HostConfig.NetworkMode
-		if container.Config.NetworkDisabled || mode.IsContainer() {
```
By moving the functionality into `UpdateNetworkSettings`, this is a case where we must handle properly.
Else, we will end up processing the allocateNetwork for `NetworkDisabled || IsContainer()` case, which is incorrect.

coolljt0725 commented:
@mavenugo  Thanks ,updated

thaJeztah commented:
@coolljt0725 looks like it fails to build :cry:

coolljt0725 commented:
@thaJeztah  Forgot to build on local before pushing to github....

cpuguy83 commented on daemon/container_operations_unix.go:
```diff
@@ -529,6 +529,29 @@ func (daemon *Daemon) updateNetwork(container *container.Container) error {
 }
 
+// UpdateNetworkSettings update the network settings
+func (daemon *Daemon) UpdateNetworkSettings(container *container.Container) error {
```
Does this need to be exported?

mavenugo replied:
good catch. it should not be exported.

coolljt0725 replied:
There is already a method `updateNetworkSettings`, so I use `UpdateNetworkSettings` at the very start. what about `updateContainerNetworkSettings`?

mavenugo commented on daemon/create.go:
```diff
@@ -106,6 +106,10 @@ func (daemon *Daemon) create(params *ContainerCreateConfig) (retC *container.Con
 		return nil, err
 	}
 
+	if err := daemon.UpdateNetworkSettings(container); err != nil {
```
This method must be called only the first time and that is the reason why it was originally protected with
`if len(container.NetworkSettings.Networks) == 0 {` in the allocateNetwork function.

Without that protection, a container restarted with multiple networks will start to behave incorrectly.

Maybe we need a test for that as well.

coolljt0725 committed:
update network settings on container creating

To make docker inspect return a consistent result of networksettings
for created container and stopped container, it's bettew to update
the network settings on container creating.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

coolljt0725 commented:
@mavenugo  Add a test, is that what you want?

mavenugo commented on daemon/create.go:
```diff
@@ -106,6 +106,10 @@ func (daemon *Daemon) create(params *ContainerCreateConfig) (retC *container.Con
 		return nil, err
 	}
 
+	if err := daemon.updateContainerNetworkSettings(container); err != nil {
```
This method must be called only the first time and that is the reason why it was originally protected with
`if len(container.NetworkSettings.Networks) == 0 {` 
in the allocateNetwork function.

Without that protection, a container restarted with multiple networks will start to behave incorrectly (especially ungraceful daemon restart cases). I guess we need to add the above protection (``if len(container.NetworkSettings.Networks) == 0 {` `) here as well.

coolljt0725 replied:
@mavenugo  This method does be called only the first time, because `create` method only be called once. Once the container has been created, the method will never been called again.

mavenugo replied:
@coolljt0725 you are right :). confused myself with create & start.

mavenugo commented:
Thanks @coolljt0725 
LGTM.

LK4D4 commented:
Makes sense. Thanks @coolljt0725 
LGTM

thaJeztah commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -891,3 +891,22 @@ func (s *DockerNetworkSuite) TestDockerNetworkConnectWithMac(c *check.C) {
+	c.Assert(networks, checker.Contains, "bridge", check.Commentf("Should return 'bridge' network"))
+}
+
+func (s *DockerNetworkSuite) TestDockerNetworkRestartWithMulipleNetworks(c *check.C) {
```
oh, just noticed a typo here; TestDockerNetworkRestartWithMulipleNetworks (muliple -> multiple )

coolljt0725 replied:
@thaJeztah  Thanks.. Fix this in #19291
